apps/OboeTester/app/src/main/java/com/mobileer/audio_device/CommunicationDeviceSpinner.java

```diff
@@ -0,0 +1,126 @@
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr){
+        super(context, attrs, defStyleAttr);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr, int mode){
+        super(context, attrs, defStyleAttr, mode);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr,
+                                      int defStyleRes, int mode){
+        super(context, attrs, defStyleAttr, defStyleRes, mode);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr,
+                                      int defStyleRes, int mode, Theme popupTheme){
+        super(context, attrs, defStyleAttr, defStyleRes, mode, popupTheme);
+        setup(context);
+    }
+
+    public AudioDeviceInfo[] getCommunicationsDevices() {
+        return mCommDeviceArray;
+    }
+
+    private void setup(Context context){
+        mContext = context;
+
+        mAudioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+
+        mDeviceAdapter = new AudioDeviceAdapter(context);
+        setAdapter(mDeviceAdapter);
+
+        // Add a default entry to the list and select it
+        mDeviceAdapter.add(new AudioDeviceListEntry(CLEAR_DEVICE_ID,
+                mContext.getString(R.string.auto_select)));
+        setSelection(0);
+        setupCommunicationDeviceListener();
+    }
+
+    @TargetApi(31)
+    private void setupCommunicationDeviceListener(){
+        // Note that we will immediately receive a call to onDevicesAdded with the list of
+        // devices which are currently connected.
+        mAudioManager.registerAudioDeviceCallback(new AudioDeviceCallback() {
+            @Override
+            public void onAudioDevicesAdded(AudioDeviceInfo[] addedDevices) {
+                updateDeviceList();
+            }
+
+            public void onAudioDevicesRemoved(AudioDeviceInfo[] removedDevices) {
+                updateDeviceList();
+                setSelection(0);
+            }
+
+            private void updateDeviceList() {
+                mDeviceAdapter.clear();
+                mDeviceAdapter.add(new AudioDeviceListEntry(CLEAR_DEVICE_ID,
+                        mContext.getString(R.string.clear)));
+                setSelection(0);
+                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
+                    List<AudioDeviceInfo> commDeviceList = mAudioManager.getAvailableCommunicationDevices();
+                    mCommDeviceArray = commDeviceList.toArray(new AudioDeviceInfo[0]);
+                    List<AudioDeviceListEntry> deviceList =
+                            AudioDeviceListEntry.createListFrom(
+                                    mCommDeviceArray, AudioManager.GET_DEVICES_OUTPUTS);
```

Comment:
What about input devices? Do we assume that communication devices have both input/output?

Reply:
The Java docs are not very clear but CommunicationsDevices are always OUTPUT. The input is chosen based on the output device. I added a comment.